package/MDAnalysis/coordinates/H5MD.py

```diff
@@ -727,3 +740,89 @@ def has_forces(self):
     @has_forces.setter
     def has_forces(self, value: bool):
         self._has['force'] = value
+
+    def __getstate__(self):
+        state = self.__dict__.copy()
+        del state['_particle_group']
+        return state
+
+    def __setstate__(self, state):
+        self.__dict__ = state
+        self._particle_group = self._file['particles'][
+                               list(self._file['particles'])[0]]
+        self[self.ts.frame]
+
+
+class H5PYPicklable(h5py.File):
+    """H5PY file object (read-only) that can be pickled.
+
+    This class provides a file-like object (as returned by
+    :class:`h5py.File`) that,
+    unlike standard Python file objects,
+    can be pickled. Only read mode is supported.
+
+    When the file is pickled, filename, mode, driver, and comm of
+    :class:`h5py.File` in the file are saved. On unpickling, the file
+    is opened by filename, mode, driver, and comm (driver and comm are
+    optional arguments). This means that for a successful unpickle, the
+    original file still has to be accessible with its filename.
+
+    Parameters
+    ----------
+    filename : str or file-like
+        a filename given a text or byte string.
+    driver : str (optional)
+        H5PY file driver used to open H5MD file
+    comm : :class:`MPI.Comm` (optional)
+        MPI communicator used to open H5MD file
+        Must be passed with `'mpio'` file driver
+
+    Example
+    -------
+    ::
+
+        f = H5PYPicklable('filename', 'r')
+        print(f['particles/trajectory/position/value'][0])
+        f.close()
+
+    can also be used as context manager::
+
+        with H5PYPicklable('filename', 'r'):
+            print(f['particles/trajectory/position/value'][0])
+
+    See Also
+    ---------
+    :class:`MDAnalysis.lib.picklable_file_io.FileIOPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.BufferIOPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.TextIOPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.GzipPicklable`
+    :class:`MDAnalysis.lib.picklable_file_io.BZ2Picklable`
+
+
+    .. versionadded:: 2.0.0
+    """
+    def __getstate__(self):
+        try:
+            driver = self.driver
+        except AttributeError:
+            driver = None
+
+        return {'name': self.filename,
+                'mode': self.mode,
+                'driver': driver}
+
+    def __setstate__(self, state):
+        if state['driver'] == 'mpio':
+            self.__init__(name=state['name'],
+                          mode=state['mode'],
+                          driver=state['driver'],
+                          comm=state['comm'])
```

Comment:
Where will `state['comm']` come from?

Reply:
I thought it was taking the `state` dict as argument from https://github.com/MDAnalysis/mdanalysis/blob/c7087b8cb1473b98ca943057d1697b1a3f91a46d/package/MDAnalysis/coordinates/H5MD.py#L744-L747

Reply:
Maybe I don't understand how pickling works but I assumed that the goal is to make each object picklable by itself, i.e., ,`H5PYPicklableFile` **by itself** can be pickled. That's the first thing to check (and add a test for!!). Once this works, `H5MDReader` can just treat it as one item in its own class dict, as it would a numpy array or a basic Python object. It would be bad program design if `H5PYPicklableFile` needed the help of `H5MDReader` because then you could never use it independently from `H5MDReader`.

Reply:
I see, that makes sense, thanks for the clarification. I'm doing some testing with `comm` on the parallel build I have on spudda